generator/generator_tests/osm_type_test.cpp

```diff
@@ -770,53 +770,122 @@ UNIT_CLASS_TEST(TestWithClassificator, OsmType_Hwtag)
 
-  TestSurfaceTypes("paved", "intermediate", "", "paved_good");
-  TestSurfaceTypes("", "intermediate", "", "unpaved_good");
-
+  TestSurfaceTypes("paved", "", "", "paved_good");
   TestSurfaceTypes("paved", "", "2", "paved_good");
+  TestSurfaceTypes("paved", "intermediate", "", "paved_bad");
   TestSurfaceTypes("", "excellent", "", "paved_good");
-  TestSurfaceTypes("wood", "", "", "paved_bad");
+  TestSurfaceTypes("", "intermediate", "", "paved_bad");
   TestSurfaceTypes("wood", "good", "", "paved_good");
   TestSurfaceTypes("wood", "", "3", "paved_good");
-  TestSurfaceTypes("pebblestone", "", "4", "unpaved_good");
-  TestSurfaceTypes("unpaved", "", "", "unpaved_bad");
-  TestSurfaceTypes("mud", "", "", "unpaved_bad");
+  TestSurfaceTypes("wood", "", "", "paved_bad");
 
-  /// @todo Is it better paved_bad?
-  TestSurfaceTypes("", "bad", "", "unpaved_good");
+  TestSurfaceTypes("pebblestone", "", "4", "unpaved_good");
+  TestSurfaceTypes("pebblestone", "", "", "unpaved_good");
+  TestSurfaceTypes("unpaved", "", "2", "unpaved_good");
+  TestSurfaceTypes("unpaved", "", "", "unpaved_good");
+  TestSurfaceTypes("unpaved", "intermediate", "", "unpaved_bad");
+  TestSurfaceTypes("unpaved", "bad", "", "unpaved_bad");
 
-  TestSurfaceTypes("", "horrible", "", "unpaved_bad");
+  TestSurfaceTypes("ground", "good", "2", "unpaved_good");
+  TestSurfaceTypes("ground", "", "5", "unpaved_good");
+  TestSurfaceTypes("ground", "", "3", "unpaved_good");
+  TestSurfaceTypes("ground", "", "2.5", "unpaved_bad");
+  TestSurfaceTypes("ground", "", "", "unpaved_bad");
   TestSurfaceTypes("ground", "", "1", "unpaved_bad");
+  TestSurfaceTypes("ground", "intermediate", "", "unpaved_bad");
+  TestSurfaceTypes("ground", "bad", "", "unpaved_bad");
+  TestSurfaceTypes("mud", "good", "1", "unpaved_good");
   TestSurfaceTypes("mud", "", "3", "unpaved_good");
-  TestSurfaceTypes("ground", "", "5", "unpaved_good");
+  TestSurfaceTypes("mud", "", "", "unpaved_bad");
+
+  TestSurfaceTypes("", "bad", "", "paved_bad");
+  TestSurfaceTypes("", "unknown", "", "paved_bad");
+  TestSurfaceTypes("", "horrible", "", "unpaved_bad");
   TestSurfaceTypes("unknown", "", "", "unpaved_good");
-  TestSurfaceTypes("", "unknown", "", "unpaved_good");
+  TestSurfaceTypes("unknown", "unknown", "", "unpaved_good");
 
   TestSurfaceTypes("asphalt;concrete", "", "", "paved_good");
   TestSurfaceTypes("concrete:plates", "", "", "paved_good");
   TestSurfaceTypes("cobblestone:flattened", "", "", "paved_bad");
   TestSurfaceTypes("dirt/sand", "", "", "unpaved_bad");
+
+  {
+    Tags const tags = {
+        {"highway", "trunk"},
+        {"smoothness", "intermediate"},
+    };
+
+    auto const params = GetFeatureBuilderParams(tags);
+
+    TEST_EQUAL(params.m_types.size(), 1, (params));
+    TEST(params.IsTypeExist(GetType({"highway", "trunk"})), (params));
+  }
+
+  {
+    Tags const tags = {
+        {"highway", "motorway"},
+        {"smoothness", "intermediate"},
```

Comment:
why smoothness is ignored for trunks & motorways?
there could be even unpaved ones

Reply:
"surface" tag is not ignored, but smoothness is controversial in the examples I saw. There are a lot of them in Ukraine, where somebody tagged all main roads like smoothness=intermediate and the app prefers nearby primary/secondary roads without tags, which are definitely not better than trunk/motorway.

Reply:
This sounds a bit like mismapping and OM helping to legalise it :)
I think its fine if OM's behavior is consistent with other routers. Is it?

And in any case add a comment describing a reason for that exception please.

Reply:
OSM routers work fine here. Mapping also satisfies OSM rules, I think. The big problem here (and we have TODOs and issues) is that we try to fit all OSM surface and smoothness tags into 4 OM types:
```
/// @todo Should make some compare constrains (like in CarModel_TrackVsGravelTertiary test)
/// to better fit these factors with reality. I have no idea, how they were set.
VehicleModel::SurfaceInitList const kCarSurface = {
  // {{surfaceType, surfaceType}, {weightFactor, etaFactor}}
  {{"psurface", "paved_good"}, {1.0, 1.0}},
  {{"psurface", "paved_bad"}, {0.6, 0.7}},
  {{"psurface", "unpaved_good"}, {0.4, 0.7}},
  {{"psurface", "unpaved_bad"}, {0.2, 0.3}}
};
```